Doc 📓: Add detailed explanation about writing new hardware interface. (#615)
ros-controls/ros2_control: hardware_interface/doc/writing_new_hardware_interface.rst

> .. _writing_new_controllers:

Author: ```suggestion
.. _writing_new_hardware_interface:
```

   For further explanation of hardware-lifecycle check the `pending pull request <https://github.com/ros-controls/ros2_control/pull/559/files#diff-2bd171d85b028c1b15b03b27d4e6dcbb87e52f705042bf111840e7a28ab268fc>`_ and for exact definitions of methods check the ``"hardware_interface/$interface_type$_interface.hpp"`` header or `doxygen documentation <http://control.ros.org/api/namespacehardware__interface.html>`_ for *Actuator*, *Sensor* or *System*.

4. **Adding definitions into source file (.cpp)**

   1. Include the header file of your hardware interface and add a namespace definition to simplify further development.

   2. Implement ``on_init`` method. There, you should all initialize member variables and process hardware's parameters from ``info`` argument.
      In the first line usually the parents ``on_init`` is called to process standard values, like name. This is done using: ``hardware_interface::(Actuator|Sensor|System)Interface::on_init(info)``.
      If all required parameters are set and valid and everything works fine return ``CallbackReturn::SUCCESS`` or ``return CallbackReturn::ERROR`` otherwise.

>    4. Write the ``on_configure`` method where you usually setup the communication to the hardware and set everything up so that the hardware can be activated.

Author: What do you think about this?

```suggestion
   4. Write the ``on_configure`` method where you usually setup the communication to the hardware, allocate the memory and set everything up so that the hardware can be activated.
```

>    5. Implement ``on_cleanup`` method, which does the opposite of ``on_configure``.

Author: ```suggestion
   5. Implement ``on_cleanup`` method, which does the opposite of ``on_configure``. Deallocate all dynamically allocated structures, i.e, anyting created with `new`, `make_shared`, `make_unique`, etc.
```

>    6. Implement ``export_state_interfaces`` and ``export_command_interfaces`` where interfaces that hardware if offering are defined.

Reviewer: please rework this sentence

>    6. Implement ``export_state_interfaces`` and ``export_command_interfaces`` where interfaces that hardware if offering are defined.

Author: Better?

```suggestion
   6. Implement ``export_state_interfaces`` and ``export_command_interfaces`` methods where interfaces that hardware if offering are defined.
```

      For the ``Sensor``-type hardware interface there is no ``export_command_interfaces`` method.

>       As a reminder, the full interface names have structure ``<joint_name>/<interface_type>``.

Reviewer: this is a little sketchy, if I don't get it immediately, others will have a hard time

Author: Do you have any proposal.

   7. (optional) For *Actuator* and *System* types of hardware interface implement ``prepare_command_mode_switch`` and ``perform_command_mode_switch`` if your hardware accepts multiple control modes.

   8. Implement the ``on_activate`` method where hardware "power" is enabled.

   9. Implement the ``on_deactivate`` method, which does the opposite of ``on_activate``.

   10. Implement ``on_shutdown`` method where hardware is shutdown gracefully.

>    11. Implement ``on_error`` method where different errors from all states are handled.

Reviewer: we could add something about recovering from errors being possible here?

Author: Probably we should link the explanation from [here](http://control.ros.org/ros2_control/hardware_interface/doc/hardware_components_userdoc.html#handling-of-errors-that-happen-during-read-and-write-calls). What do you think.

   14. IMPORTANT: At the end of your file after the namespace is closed, add the ``PLUGINLIB_EXPORT_CLASS`` macro.
      For this you will need to include the ``"pluginlib/class_list_macros.hpp"`` header.
      As first parameters you should provide exact hardware interface class, e.g., ``<my_hardware_interface_package>::<RobotHardwareInterfaceName>``, and as second the base class, i.e., ``hardware_interface::(Actuator|Sensor|System)Interface``.

5. **Writing export definition for pluginlib**

   1. Create the ``<my_hardware_interface_package>.xml`` file in the package and add a definition of the library and hardware interface's class which has to be visible for the pluginlib.
      The easiest way to do that is to check definition for fake components in the `ros2_control <https://github.com/ros-controls/ros2_control/blob/master/hardware_interface/fake_components_plugin_description.xml>`_ package.

   2. Usually, the plugin name is defined by the package (namespace) and the class name, e.g.,
      ``<my_hardware_interface_package>/<RobotHardwareInterfaceName>``.
      This name defines the hardware interface's type when the resource manager searches for it.

>       The other two parameters have to correspond to the definition done in macro at the bottom of the ``<robot_hardware_interface_name>.cpp`` file.

Reviewer: ```suggestion
      The other two parameters have to correspond to the definition done in the macro at the bottom of the ``<robot_hardware_interface_name>.cpp`` file.
```

> 6. **Writing simple test to check if the controller can be found and loaded**

Reviewer: ```suggestion
6. **Writing a simple test to check if the controller can be found and loaded**
```

   4. Add ament dependencies needed by the library. You should add at least those listed under 1.

   5. Export for pluginlib description file using the following command:
      .. code:: cmake

         pluginlib_export_plugin_description_file(hardware_interface <my_hardware_interface_package>.xml)

   6. Add install directives for targets and include directory.

   7. In the test section add the following dependencies: ``ament_cmake_gmock``, ``hardware_interface``.

   8. Add compile definitions for the tests using the ``ament_add_gmock`` directive.
      For details, see how it is done for fake hardware in the `ros2_control <https://github.com/ros-controls/ros2_control/blob/master/hardware_interface/CMakeLists.txt>`_ package.

   9. (optional) Add your hardware interface`s library into ``ament_export_libraries`` before ``ament_package()``.

8. **Add dependencies into ``package.xml`` file**

   1. Add at least the following packages into ``<depend>`` tag: ``hardware_interface``, ``pluginlib``, ``rclcpp``, and ``rclcpp_lifecycle``.

   2. Add at least the following package into ``<test_depend>`` tag: ``ament_add_gmock`` and ``hardware_interface``.

> 9. **Compiling and testing the controller**

Reviewer: ```suggestion
9. **Compiling and testing the hardware component**
```

>    1. Now everything is ready to compile the controller using the ``colcon build <my_hardware_interface_package>`` command.

Reviewer: ```suggestion
   1. Now everything is ready to compile the hardware component using the ``colcon build <my_hardware_interface_package>`` command.
```

      Remember to go into the root of your workspace before executing this command.

   2. If compilation was successful, source the ``setup.bash`` file from the install folder and execute ``colcon test <my_hardware_interface_package>`` to check if the new controller can be found through ``pluginlib`` library and be loaded by the controller manager.

That's it! Enjoy writing great controllers!

Useful External References
---------------------------

> - `Templates and scripts for generating controllers shell <https://stoglrobotics.github.io/ros_team_workspace/use-cases/ros2_control/setup_robot_hardware_interface.html>`_ **NOTE**: The script is not up-to date with the new API begging from galactic.

Reviewer: ```suggestion
- `Templates and scripts for generating controllers shell <https://stoglrobotics.github.io/ros_team_workspace/use-cases/ros2_control/setup_robot_hardware_interface.html>`_ **NOTE**: The script is currently only recommended to use for Foxy, not compatible with the API from Galactic and onwards.
```
